src/runtime/pipeline/pipeline_executor.cc

```diff
@@ -0,0 +1,169 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*!
+ * \file pipeline_executor.cc
+ */
+#include "pipeline_executor.h"
+
+#include <tvm/runtime/registry.h>
+
+namespace tvm {
+namespace runtime {
+
+/*!
+ *\bief Stop pipeline run.
+ */
+void SubGraphRuntime::Stop() { pipeline_stop(runtimes); }
+/*!
+ * \brief Run all the operations one by one.
+ */
+void SubGraphRuntime::Run() { pipeline_run(runtimes); }
+
+void SubGraphRuntime::Init(const Array<tvm::runtime::Module>& modules,
+                           const std::string& pipeline_json) {
+  std::istringstream is(pipeline_json);
+  dmlc::JSONReader reader(&is);
+  this->Load(&reader);
+  pipeline_init(modules, &runtimes, &pipeline_conf);
+  return;
+}
+
+/*!
+ * \brief set index-th input to the graph.
+ * \param index The input index.
+ * \param data_in The input data.
+ */
+void SubGraphRuntime::SetInput(int index, DLTensor* data_in) {
+  auto gruntime = runtimes.front();
+  gruntime->runtimePtr->SetInput(index, data_in);
+}
+
+void SubGraphRuntime::SetInput(const std::string& name, DLTensor* data_in) {
+  auto gruntime = runtimes.front();
+  gruntime->runtimePtr->SetInput(name, data_in);
+}
+
+/*!
+ * \brief Get the number of outputs
+ *
+ * \return The number of outputs from last pipeline.
+ */
+int SubGraphRuntime::NumOutputs() const { return runtimes.back()->runtimePtr->NumOutputs(); }
+
+/*!
+ * \brief Get the number of inputs
+ *
+ * \return The number of inputs to the first pipeline.
+ */
+int SubGraphRuntime::NumInputs() const { return runtimes.front()->runtimePtr->NumInputs(); }
```

Comment:
I don't think this logic is correct? Inputs can come from non-front modules and outputs can come from non-back modules. For example, if users don't bind the weights and choose to set them on the fly, then every module must have inputs.

Reply:
fixed.